eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs

```diff
@@ -0,0 +1,255 @@
+  | DelegCertDRepNotRegisteredDELEG
+      !(Credential 'Voting (EraCrypto era)) -- Credential which is not registered
+  | DelegCertStakeVoteDRepNotRegisteredDELEG
+      !(Credential 'Voting (EraCrypto era)) -- Credential which is not registered
+
+  | RegDelegCertProvidedCoinNotEqualToExpectedPoolDepositDELEG
+      !Coin
+  | RegDelegCertStakeKeyAlreadyRegisteredDELEG
+      !(Credential 'Staking (EraCrypto era)) -- Credential which is already registered
+  | RegDelegCertDRepNotRegisteredDELEG
+      !(Credential 'Voting (EraCrypto era)) -- Credential which is not registered
+  | RegDelegCertStakeVoteDRepNotRegisteredDELEG
+      !(Credential 'Voting (EraCrypto era)) -- Credential which is not registered
+
+  | RegDelegCertDRepAlreadyRegisteredDELEG
+      !(Credential 'Staking (EraCrypto era)) -- Credential which is already registered
+  | WrongCertificateTypeDELEG -- The DCertPool constructor should not be used by this transition
+  deriving (Generic)
+
+instance 
+  NoThunks (PredicateFailure (EraRule "DELEG" era)) => 
+  NoThunks (ConwayDelegPredFailure era)
+deriving instance
+  Eq (PredicateFailure (EraRule "DELEG" era)) =>
+  Eq (ConwayDelegPredFailure era)
+deriving instance
+  Show (PredicateFailure (EraRule "DELEG" era)) =>
+  Show (ConwayDelegPredFailure era)
+
+-- newtype ConwayDelegEvent era = NewEpoch EpochNo
+newtype ConwayDelegEvent era = DelegEvent (Event (EraRule "DELEG" era))
+
+instance 
+  ( EraTx era
+  , ConwayEraTxBody era
+  , State (EraRule "DELEG" era) ~ CertState era
+  , Signal (EraRule "DELEG" era) ~ TxCert era
+  , TxCert era ~ ConwayTxCert era
+  , Environment (EraRule "DELEG" era) ~ DelegEnv era
+  , EraRule "DELEG" era ~ ConwayDELEG era
+  ) => 
+  STS (ConwayDELEG era) where
+  type State (ConwayDELEG era) = CertState era
+  type Signal (ConwayDELEG era) = TxCert era
+  type Environment (ConwayDELEG era) = DelegEnv era
+  type BaseM (ConwayDELEG era) = ShelleyBase
+  type PredicateFailure (ConwayDELEG era) = ConwayDelegPredFailure era
+  type Event (ConwayDELEG era) = ConwayDelegEvent era
+
+  transitionRules = [conwayDelegTransition @era]
+
+conwayDelegTransition :: 
+  forall era.
+  ( EraTx era
+  -- , ConwayEraTxBody era
+  -- , ShelleyEraTxBody era
+  -- , State (EraRule "DELEG" era) ~ CertState era
+  -- , Signal (EraRule "DELEG" era) ~ TxCert era
+  , TxCert era ~ ConwayTxCert era
+  -- , Environment (EraRule "DELEG" era) ~ DelegEnv era
+  -- , EraRule "DELEG" era ~ ConwayDELEG era
+  ) => 
+  TransitionRule (ConwayDELEG era)
+conwayDelegTransition = do
+  TRC 
+    ( DelegEnv _slot ptr _acnt pp
+    , cState@CertState
+        { certVState = VState{vsDReps}
+        , certPState = _
+        , certDState = certDState@DState{dsUnified} 
+        } 
+    , c
+    ) <- judgmentContext
+  let pd = pp ^. ppKeyDepositL
+  case c of
+    ConwayTxCertDeleg (ConwayRegCert stakeCred sMayCoin) -> do
+      sMayCoin == SNothing || sMayCoin == SJust pd
```

Comment:
```suggestion
      sMayCoin == SJust pd
```
I think for now we'll want to allow `SNothing` in the deposits so we can make the decoder backwards compatible. The deposit field should be optional in Conway.

Reply:
You meant to write `/=`, right? :grin:

Reply:
yes, my bad

Reply:
It is indeed similar to an `unless` :+1: